overlord/patch/patch6_2.go

```diff
@@ -0,0 +1,155 @@
+	Amend            bool `json:"amend,omitempty"`
+	IsAutoRefresh    bool `json:"is-auto-refresh,omitempty"`
+	NoReRefresh      bool `json:"no-rerefresh,omitempty"`
+	RequireTypeBase  bool `json:"require-base-type,omitempty"`
+}
+
+type patch62SnapState struct {
+	SnapType string             `json:"type"`
+	Sequence []*patch62SideInfo `json:"sequence"`
+	Active   bool               `json:"active,omitempty"`
+	Current  snap.Revision      `json:"current"`
+	Channel  string             `json:"channel,omitempty"`
+	patch62Flags
+	Aliases              interface{} `json:"aliases,omitempty"`
+	AutoAliasesDisabled  bool        `json:"auto-aliases-disabled,omitempty"`
+	AliasesPending       bool        `json:"aliases-pending,omitempty"`
+	UserID               int         `json:"user-id,omitempty"`
+	InstanceKey          string      `json:"instance-key,omitempty"`
+	CohortKey            string      `json:"cohort-key,omitempty"`
+	RefreshInhibitedTime *time.Time  `json:"refresh-inhibited-time,omitempty"`
+}
+
+type patch62SnapSetup struct {
+	Channel   string    `json:"channel,omitempty"`
+	UserID    int       `json:"user-id,omitempty"`
+	Base      string    `json:"base,omitempty"`
+	Type      snap.Type `json:"type,omitempty"`
+	PlugsOnly bool      `json:"plugs-only,omitempty"`
+	CohortKey string    `json:"cohort-key,omitempty"`
+	Prereq    []string  `json:"prereq,omitempty"`
+	patch62Flags
+	SnapPath     string           `json:"snap-path,omitempty"`
+	DownloadInfo interface{}      `json:"download-info,omitempty"`
+	SideInfo     *patch62SideInfo `json:"side-info,omitempty"`
+	patch62auxStoreInfo
+	InstanceKey string `json:"instance-key,omitempty"`
+}
+
+type patch62auxStoreInfo struct {
+	Media interface{} `json:"media,omitempty"`
+}
+
+// patch6_2:
+//  - ensure snapd snaps in the snapstate have TypeSnapd for backward compatibility with old snapd snap releases.
+//  - ensure snapd snaps have TypeSnapd in pending install tasks.
+func patch6_2(st *state.State) error {
+	var snaps map[string]*json.RawMessage
+	if err := st.Get("snaps", &snaps); err != nil && err != state.ErrNoState {
+		return fmt.Errorf("internal error: cannot get snaps: %s", err)
+	}
+
+	// migrate snapstate
+	for name, raw := range snaps {
+		var snapst patch62SnapState
+		if err := json.Unmarshal([]byte(*raw), &snapst); err != nil {
+			return err
+		}
+
+		if snapst.SnapType == string(snap.TypeSnapd) {
+			// we can have just one snapd snap
+			break
+		}
+		if snapst.SnapType == string(snap.TypeApp) {
+			for _, seq := range snapst.Sequence {
+				if snap.SnapIDSnapd(seq.SnapID) {
+					snapst.SnapType = string(snap.TypeSnapd)
+					data, err := json.Marshal(snapst)
+					if err != nil {
+						return err
+					}
+					newRaw := json.RawMessage(data)
+					snaps[name] = &newRaw
+					st.Set("snaps", snaps)
+					// we can have just one snapd snap
+					break
+				}
+			}
+		}
+	}
+
+	// migrate tasks' snap setup
+	for _, task := range st.Tasks() {
+		if task.Status().Ready() {
```

Comment:
because of the fact that some tasks refer to previous tasks for their snap-setup I fear we need to check for full Change readiness and not only task by task

Reply:
That's a good point, you're right, will fix, thanks!